File: langs/typescript/src/rpc-retry.js
/** Deterministic RPC retry planning; transport, clocks, randomness and credentials stay outside. */
export const RETRY_REASON = Object.freeze({
  retry: 1, invalid_input: 2, cancelled: 3, unsafe_replay: 4,
  attempts_exhausted: 5, deadline_exhausted: 6, non_retryable: 7,
});

const POLICY_KEYS = Object.freeze([
  'max_attempts', 'timeout_ms', 'initial_backoff_ms', 'max_backoff_ms',
]);
const ATTEMPT_KEYS = Object.freeze([
  'attempts_completed', 'elapsed_ms', 'code', 'cancelled', 'replay_safe', 'jitter_permille',
]);
const bounded = (value, min, max) => Number.isSafeInteger(value) && value >= min && value <= max;

// Accept inert JSON-shaped records only. Inherited values and accessor properties are not inputs.
function record(value, required, optional = []) {
  if (value === null || typeof value !== 'object') return false;
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) return false;
  const descriptors = Object.getOwnPropertyDescriptors(value);
  return required.every((key) => Object.hasOwn(descriptors, key))
    && Reflect.ownKeys(descriptors).every((key) => typeof key === 'string'
      && (required.includes(key) || optional.includes(key))
      && Object.hasOwn(descriptors[key], 'value') && descriptors[key].enumerable);
}

const stop = (reason) => Object.freeze({ retry: false, delay_ms: 0, reason });

/**
 * Plan AFTER a failed completed attempt. max_attempts includes the original call.
 * replay_safe is trusted method metadata, not a request parameter or proof supplied by a client.
 * An idempotency key alone does not make a write safe to replay.
 * elapsed_ms is monotonic elapsed time; the host rechecks cancellation/deadline before sending.
 * jitter_permille is an injected integer sample in [0, 1000]; this function draws no randomness.
 * retry_after_ms is an optional SERVER MINIMUM, never shortened to the local backoff cap.
 * No request payload, token, idempotency key, SQL or driver error is accepted or returned.
 */
export function planRpcRetry(policy, attempt) {
  if (!record(policy, POLICY_KEYS) || !record(attempt, ATTEMPT_KEYS, ['retry_after_ms'])) {
    return stop(RETRY_REASON.invalid_input);
  }
  if (!bounded(policy.max_attempts, 1, 8)
      || !bounded(policy.timeout_ms, 1, 3_600_000)
      || !bounded(policy.initial_backoff_ms, 1, 60_000)
      || !bounded(policy.max_backoff_ms, policy.initial_backoff_ms, 60_000)
      || !bounded(attempt.attempts_completed, 1, 8)
      || !bounded(attempt.elapsed_ms, 0, 3_600_000)
      || !bounded(attempt.code, 0, 16)
      || typeof attempt.cancelled !== 'boolean' || typeof attempt.replay_safe !== 'boolean'
      || !bounded(attempt.jitter_permille, 0, 1000)
      || (Object.hasOwn(attempt, 'retry_after_ms') && !bounded(attempt.retry_after_ms, 0, 3_600_000))) {
    return stop(RETRY_REASON.invalid_input);
  }
  if (attempt.cancelled) return stop(RETRY_REASON.cancelled);
  if (!attempt.replay_safe) return stop(RETRY_REASON.unsafe_replay);
  if (attempt.attempts_completed >= policy.max_attempts) return stop(RETRY_REASON.attempts_exhausted);
  if (attempt.elapsed_ms >= policy.timeout_ms) return stop(RETRY_REASON.deadline_exhausted);
  // gRPC/Connect RESOURCE_EXHAUSTED (8) and UNAVAILABLE (14), explicitly allowlisted.
  if (attempt.code !== 8 && attempt.code !== 14) return stop(RETRY_REASON.non_retryable);
  const cap = Math.min(policy.max_backoff_ms,
    policy.initial_backoff_ms * (2 ** (attempt.attempts_completed - 1)));
  const jittered = Math.floor(cap * attempt.jitter_permille / 1000);
  const delay = Math.max(jittered, attempt.retry_after_ms ?? 0);
  // Equality leaves no time for the call; never shorten a server minimum to fit a deadline.
  if (delay >= policy.timeout_ms - attempt.elapsed_ms) return stop(RETRY_REASON.deadline_exhausted);
  return Object.freeze({ retry: true, delay_ms: delay, reason: RETRY_REASON.retry });
}

File: langs/dart/lib/rpc_retry.dart
/// Pure RPC retry planning. The host owns transport, time, randomness and cancellation.
class RpcRetryPolicy {
  final int maxAttempts;
  final int timeoutMs;
  final int initialBackoffMs;
  final int maxBackoffMs;
  const RpcRetryPolicy({required this.maxAttempts, required this.timeoutMs,
    required this.initialBackoffMs, required this.maxBackoffMs});
}

class RpcRetryAttempt {
  final int attemptsCompleted;
  final int elapsedMs;
  final int code;
  final bool cancelled;
  /// Trusted method metadata; a client-supplied idempotency key alone is not sufficient.
  final bool replaySafe;
  final int jitterPermille;
  /// Server minimum, never shortened to fit a deadline or the local backoff cap.
  final int? retryAfterMs;
  const RpcRetryAttempt({required this.attemptsCompleted, required this.elapsedMs,
    required this.code, required this.cancelled, required this.replaySafe,
    required this.jitterPermille, this.retryAfterMs});
}

enum RpcRetryReason {
  retry(1), invalidInput(2), cancelled(3), unsafeReplay(4),
  attemptsExhausted(5), deadlineExhausted(6), nonRetryable(7);
  final int wireCode;
  const RpcRetryReason(this.wireCode);
}

class RpcRetryDecision {
  final bool retry;
  final int delayMs;
  final RpcRetryReason reason;
  const RpcRetryDecision._(this.retry, this.delayMs, this.reason);
}

bool _bounded(int value, int min, int max) => value >= min && value <= max;
RpcRetryDecision _stop(RpcRetryReason reason) => RpcRetryDecision._(false, 0, reason);

/// Call after a failed completed attempt. maxAttempts includes the original call.
/// elapsedMs is monotonic elapsed time. Recheck cancellation/deadline before each send.
/// No request, key, token, ORM entity, driver error, environment or global state is used.
RpcRetryDecision planRpcRetry(RpcRetryPolicy policy, RpcRetryAttempt attempt) {
  final minimum = attempt.retryAfterMs;
  final valid = _bounded(policy.maxAttempts, 1, 8)
    && _bounded(policy.timeoutMs, 1, 3600000)
    && _bounded(policy.initialBackoffMs, 1, 60000)
    && _bounded(policy.maxBackoffMs, policy.initialBackoffMs, 60000)
    && _bounded(attempt.attemptsCompleted, 1, 8)
    && _bounded(attempt.elapsedMs, 0, 3600000)
    && _bounded(attempt.code, 0, 16)
    && _bounded(attempt.jitterPermille, 0, 1000)
    && (minimum == null || _bounded(minimum, 0, 3600000));
  if (!valid) return _stop(RpcRetryReason.invalidInput);
  if (attempt.cancelled) return _stop(RpcRetryReason.cancelled);
  if (!attempt.replaySafe) return _stop(RpcRetryReason.unsafeReplay);
  if (attempt.attemptsCompleted >= policy.maxAttempts) return _stop(RpcRetryReason.attemptsExhausted);
  if (attempt.elapsedMs >= policy.timeoutMs) return _stop(RpcRetryReason.deadlineExhausted);
  if (attempt.code != 8 && attempt.code != 14) return _stop(RpcRetryReason.nonRetryable);
  final exponential = policy.initialBackoffMs * (1 << (attempt.attemptsCompleted - 1));
  final cap = exponential < policy.maxBackoffMs ? exponential : policy.maxBackoffMs;
  final jittered = cap * attempt.jitterPermille ~/ 1000;
  final serverMinimum = minimum ?? 0;
  final delay = jittered > serverMinimum ? jittered : serverMinimum;
  if (delay >= policy.timeoutMs - attempt.elapsedMs) return _stop(RpcRetryReason.deadlineExhausted);
  return RpcRetryDecision._(true, delay, RpcRetryReason.retry);
}
